File: orchalanguage-spring-integration-implementation/src/test/java/orcha/lang/compiler/referenceimpl/LexicalAnalysisTest.java
package orcha.lang.compiler.referenceimpl;

import orcha.lang.compiler.*;
import orcha.lang.compiler.referenceimpl.springIntegration.SendInstructionForSpringIntegration;
import orcha.lang.compiler.referenceimpl.springIntegration.SpringIntegrationAutoConfiguration;
import orcha.lang.compiler.referenceimpl.springIntegration.WhenInstructionForSpringIntegration;
import orcha.lang.compiler.syntax.Instruction;
import orcha.lang.compiler.syntax.SendInstruction;
import orcha.lang.compiler.syntax.TitleInstruction;
import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringRunner;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@RunWith(SpringRunner.class)
@SpringBootTest
public class LexicalAnalysisTest {

    @Autowired
    LexicalAnalysis lexicalAnalysisForTest;

    @Test
    public void test() {

        try{
            Assert.assertTrue(lexicalAnalysisForTest != null);

            List<String> linesOfCode = new ArrayList<String>(
                    Arrays.asList(
                            "title: title",
                            "",
                            "receive order from customer",
                            "when \"order terminates\"",
                            "send order.number to customer"));

            OrchaProgram orchaProgram = lexicalAnalysisForTest.analysis(linesOfCode);

            Assert.assertNotNull(orchaProgram);
            Assert.assertNotNull(orchaProgram.getIntegrationGraph());
            Assert.assertNotNull(orchaProgram.getOrchaMetadata());

            OrchaMetadata orchaMetadata = orchaProgram.getOrchaMetadata();
            List<Instruction> metadata = orchaMetadata.getMetadata();
            Assert.assertNotNull(metadata);
            Assert.assertEquals(metadata.size(), 1);
            Instruction instruction = metadata.get(0);
            Assert.assertTrue(instruction instanceof TitleInstruction);


            List<IntegrationNode> graphOfInstructions = orchaProgram.getIntegrationGraph();

            Assert.assertTrue(graphOfInstructions.size() == 3);

            IntegrationNode integrationNode = graphOfInstructions.get(1);
            Assert.assertNotNull(integrationNode);
            Instruction inst = integrationNode.getInstruction();
            inst.analysis();
            Assert.assertNotNull(inst);
            Assert.assertTrue(inst instanceof WhenInstructionForSpringIntegration);
            WhenInstructionForSpringIntegration whenInstruction = (WhenInstructionForSpringIntegration)inst;
            Assert.assertEquals(whenInstruction.getLineNumber(), 4);
            Assert.assertTrue(whenInstruction.getCommand().equals("when"));

            integrationNode = graphOfInstructions.get(2);
            Assert.assertNotNull(integrationNode);
            inst = integrationNode.getInstruction();
            Assert.assertNotNull(inst);
            inst.analysis();
            Assert.assertTrue(inst instanceof SendInstructionForSpringIntegration);
            SendInstructionForSpringIntegration sendInstruction = (SendInstructionForSpringIntegration)inst;
            Assert.assertEquals(sendInstruction.getLineNumber(), 5);
            Assert.assertTrue(sendInstruction.getCommand().equals("send"));
            Assert.assertEquals(sendInstruction.getVariables(), "payload.number");

        }catch(OrchaCompilationException e){
            Assert.fail(e.getMessage());
        }
    }

}

File: orchalang-spring-integration-implementation/src/test/kotlin/orcha/lang/compiler/referenceimpl/LexicalAnalysisTest.kt
package orcha.lang.compiler.referenceimpl

import orcha.lang.compiler.*
import orcha.lang.compiler.referenceimpl.springIntegration.*
import orcha.lang.compiler.syntax.SendInstruction
import orcha.lang.compiler.syntax.TitleInstruction
import orcha.lang.compiler.syntax.WhenInstruction
import orcha.lang.configuration.*
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.TestConfiguration
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.DependsOn
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit4.SpringRunner
import java.util.*

@RunWith(SpringRunner::class)
//@SpringBootTest
@ContextConfiguration(classes = [LexicalAnalysisTest.CompilerReferenceImplTestConfiguration::class])
class LexicalAnalysisTest {

    @TestConfiguration
    internal class CompilerReferenceImplTestConfiguration {

        @Bean
        @DependsOn("whenInstruction", "sendInstruction")
        internal fun lexicalAnalysisForTest(): LexicalAnalysis {
            return LexicalAnalysisImpl()
        }

        @Bean(name = ["whenInstruction"])
        fun whenInstructionFactory(): WhenInstructionFactory {
            return WhenInstructionFactory()
        }

        @Bean
        @Throws(Exception::class)
        fun whenInstruction(): WhenInstruction? {
            return whenInstructionFactory().getObject()
        }

        @Bean(name = ["sendInstruction"])
        fun sendInstructionFactory(): SendInstructionFactory {
            return SendInstructionFactory()
        }

        @Bean
        @Throws(Exception::class)
        fun sendInstruction(): SendInstruction? {
            return sendInstructionFactory().getObject()
        }

    }


    @Autowired
    internal var lexicalAnalysisForTest: LexicalAnalysis? = null

    @Test
    fun test() {

        try {
            Assert.assertTrue(lexicalAnalysisForTest != null)

            val linesOfCode = ArrayList(
                    Arrays.asList(
                            "title: title",
                            "",
                            "receive order from customer",
                            "when \"order terminates\"",
                            "send order.number to customer"))

            val orchaProgram = lexicalAnalysisForTest!!.analysis(linesOfCode)

            Assert.assertNotNull(orchaProgram)
            Assert.assertNotNull(orchaProgram.integrationGraph)
            Assert.assertNotNull(orchaProgram.orchaMetadata)

            val orchaMetadata = orchaProgram.orchaMetadata
            val metadata = orchaMetadata!!.metadata
            Assert.assertNotNull(metadata)
            Assert.assertEquals(metadata.size.toLong(), 1)
            val instruction = metadata[0]
            Assert.assertTrue(instruction is TitleInstruction)


            val graphOfInstructions = orchaProgram.integrationGraph

            Assert.assertTrue(graphOfInstructions!!.size == 3)

            var integrationNode = graphOfInstructions!![1]
            Assert.assertNotNull(integrationNode)
            var inst = integrationNode.instruction
            if (inst != null) {
                inst.analysis()
            }
            Assert.assertNotNull(inst)
            Assert.assertTrue(inst is WhenInstructionForSpringIntegration)
            val whenInstruction = inst as WhenInstructionForSpringIntegration
            Assert.assertEquals(whenInstruction.getLineNumber(), 4)
            Assert.assertTrue(whenInstruction.getCommand().equals("when"))

            integrationNode = graphOfInstructions[2]
            Assert.assertNotNull(integrationNode)
            inst = integrationNode.instruction
            Assert.assertNotNull(inst)
            inst?.analysis()
            Assert.assertTrue(inst is SendInstructionForSpringIntegration)
            val sendInstruction = inst as SendInstructionForSpringIntegration
            Assert.assertEquals(sendInstruction.getLineNumber(), 5)
            Assert.assertTrue(sendInstruction.getCommand().equals("send"))
            Assert.assertEquals(sendInstruction.getVariables(), "payload.number")

        } catch (e: OrchaCompilationException) {
            Assert.fail(e.message)
        }

    }

}
